stdlib/public/runtime/LibPrespecialized.cpp

```diff
@@ -322,13 +379,117 @@ swift::getLibPrespecializedMetadata(const TypeContextDescriptor *description,
   }
 
   auto key = mangling.result();
-  auto *metadataMap = data->getMetadataMap();
+  auto *metadataMap = state.data->getMetadataMap();
   auto *element = metadataMap->find(key.data(), key.size());
   auto *result = element ? element->value : nullptr;
   LOG("found %p for key '%.*s'.", result, (int)key.size(), key.data());
   return result;
 }
 
+static Metadata *
+getMetadataFromPointerKeyedMap(const LibPrespecializedState &state,
+                               const TypeContextDescriptor *description,
+                               const void *const *arguments) {
+#if DYLD_FIND_POINTER_HASH_TABLE_ENTRY_DEFINED
+  auto *generics = description->getGenericContext();
+  if (!generics)
+    return nullptr;
+
+  auto argumentCount = generics->getGenericContextHeader().NumKeyArguments;
+
+  auto *map = state.data->getPointerKeyedMetadataMap();
+  auto result = _dyld_find_pointer_hash_table_entry(map, description,
+                                                    argumentCount, arguments);
+  LOG("Looking up description %p in dyld table, found %p.", description,
+      result);
+  return (Metadata *)result;
+#else
+  LOG("Looking up description %p but dyld hash table call not available.",
+      description);
+  return nullptr;
+#endif
+}
+
+// When we have a pointer-keyed map from a debug library, it's not built as a
+// hash table. We just scan it linearly.
+static Metadata *getMetadataFromPointerKeyedMapDebugMode(
+    const LibPrespecializedState &state,
+    const TypeContextDescriptor *description, const void *const *arguments) {
+  auto *generics = description->getGenericContext();
+  if (!generics)
+    return nullptr;
+
+  auto argumentCount = generics->getGenericContextHeader().NumKeyArguments;
+  auto *mapPtr = state.data->getPointerKeyedMetadataMap();
+
+  struct MapKey {
+    size_t count;
+    void *pointers[];
+  };
+
+  struct MapEntry {
+    const MapKey *key;
+    Metadata *value;
+  };
+
+  struct Map {
+    size_t count;
+    MapEntry entries[];
+  };
+
+  const Map *map = reinterpret_cast<const Map *>(mapPtr);
+  for (size_t i = 0; i < map->count; i++) {
+    auto &entry = map->entries[i];
+
+    // Keys are descriptor followed by arguments, so their count is 1 plus the
+    // argument count.
+    if (entry.key->count != argumentCount + 1)
+      continue;
+
+    // Check the descriptor.
+    if (description != entry.key->pointers[0])
+      continue;
+
+    // Check the rest. The pointers array is now offset by 1 since index 0 is
+    // the descriptor.
+    bool equal = true;
+    for (size_t j = 0; j < argumentCount; j++)
+      if (entry.key->pointers[j + 1] != arguments[j])
+        equal = false;
```

Comment:
Could `break` here too to drop out after the first argument mismatch.

Reply:
Added.